### PR TITLE
End-to-end tests

### PR DESCRIPTION
**Pull request by vkarpov15**
Just about full E2E tests: they operate at the level of modifying Vue component state and calling Vue component methods, but those methods make HTTP requests to the backend, which then stores data in local MongoDB and uses local Oso binary. Major benefit is this E2E test is close to what happens in prod, minus actual button clicks and DOM events, but also runs in Node.js (no browser required) and is about as fast and reliable as any Node.js test. Should give us a good way of automatically confirming that Oso Golf still works when we make changes.

### CODE REVIEW

**Comment by gsarjeant** [comment]
Thanks, @vkarpov15 - these look great. Do you have some time to address the merge conflicts?

**Comment by vkarpov15** [comment]
Yeah I'm working my way through merge conflicts. There's a lot of them, I'll also need to adjust our lint configs to account for all the changes in #5

**Comment by gsarjeant** [comment]
@vkarpov15 @sploiselle  I didn't get to this before I left on PTO. I see that the merge conflicts are resolved and I went through the first couple of holes in the preview environment. It looks good to me, but I'm reviewing from my phone.

Sean, if you have some time to make sure that this is consistent with your changes from last week, could you merge it? If not, no worries, I'll merge it in when I get back.

File: backend/index.js [hunk 1]
```diff
@@ -0,0 +1,10 @@
+'use strict';
+
+exports.authorize = require('./authorize');
+exports.clearContextFacts = require('./clearContextFacts');
+exports.deleteFact = require('./deleteFact');
+exports.leaderboard = require('./leaderboard');
+exports.resumeGame = require('./resumeGame');
+exports.startGame = require('./startGame');
+exports.tell = require('./tell');
+exports.verifySolutionForLevel = require('./verifySolutionForLevel');
```
Comment (sploiselle) [inline backend/index.js]: missing newline at EOF

File: src/api.js [hunk 2]
```diff
@@ -0,0 +1,9 @@
+'use strict';
+
+const axios = require('axios');
+
+const client = typeof API_URL === 'undefined' ? axios.create() : axios.create({
+  baseURL: API_URL
+});
+
+module.exports = client;
```
Comment (sploiselle) [inline src/api.js]: newline at EOF

File: src/index.js [hunk 3]
```diff
@@ -1,15 +1,14 @@
...
-console.log(sessionId);
-window.localStorage.setItem('_gitclubGameSession', sessionId);
-
 const components = require('./components');
 const levels = require('../levels');
 const vanillatoasts = require('vanillatoasts');
 
+const sessionId = window.localStorage.getItem('_gitclubGameSession')
+  || [...Array(30)].map(() => Math.random().toString(36)[2]).join('');
+console.log(sessionId);
```
Comment (sploiselle) [inline src/index.js]: Is logging this to the console intentional?

File: src/policy.polar [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+    # reader permissions
+    "read" if "reader";
+
+    # editor permissions
+    "write" if "editor";
+}
+
+has_permission(_: Actor, "read", repo: Repository) if
+    is_public(repo, true);
+
```
Comment (sploiselle) [inline src/policy.polar:47]: nit: errant newline

File: src/policy.polar [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+has_role(actor: Actor, role: String, repo: Repository) if
+  org matches Organization and
+  has_relation(repo, "organization", org) and
+  has_default_role(org, role) and
+  has_role(actor, "member", org);
+
+has_role(user: User, role: String, resource: Resource) if
+  group matches Group and
+  has_group(user, group) and
+  has_role(group, role, resource);
```
Comment (sploiselle) [inline src/policy.polar]: newline at EOF

File: test/e2e.test.js [hunk 6]
```diff
@@ -0,0 +1,196 @@
...
+        resourceType: 'Organization',
+        resourceId: 'osohq',
+        pass: true,
+        shouldFail: true
+      }
+    ]);
+  });
+});
+
+const kebabize = (str) => str.replace(/[A-Z]+(?![a-z])|[A-Z]/g, ($, ofs) => (ofs ? '-' : '') + $.toLowerCase());
```
Comment (sploiselle) [inline test/e2e.test.js]: newline at EOF

**Review by sploiselle — APPROVED** [review]
LGTM once comments addressed

File: src/index.js (hunk 3 above)
Comment (vkarpov15) [reply src/index.js]: Removed in 27f8c59

**Pull request merged**